### PR TITLE
Reduce lock contention for frame writes

### PR DESCRIPTION
**Pull request by YulerB**
Further discussion can be found at:

https://groups.google.com/forum/?nomobile=true#!topic/rabbitmq-users/OyoYTkAyCSk

### CODE REVIEW

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 1]
```diff
@@ -1,321 +1,304 @@
...
+                    netstream = SslHelper.TcpUpgrade(netstream, endpoint.Ssl);
+                }
+                catch (Exception)
+                {
+                    Close();
+                    throw;
+                }
+            }
+            m_reader = new NetworkBinaryReader(new BufferedStream(netstream, m_socket.Client.ReceiveBufferSize));
+            m_writer = new NetworkBinaryWriter(netstream);
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Removed use of bufferedstream on the write side.
Information was being written to the stream and flushed immediately.
The change to chunk information onto the stream means we don't have to buffer.
Since we don't have to buffer, we also don't need to lock the flush.
The write doesn't need to be locked when done as a chunk.

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs [hunk 2]
```diff
@@ -994,7 +994,7 @@ protected void RecoverQueues()
                         // anything to recover. MK.
                         PropagateQueueNameChangeToBindings(oldName, newName);
                         PropagateQueueNameChangeToConsumers(oldName, newName);
-                        // see rabbitmq/rabbitmq-dotnet-client#43
+                        // see RabbitMQ/RabbitMQ-dotnet-client#43
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: Please revert this change, thanks!

File: projects/client/RabbitMQ.Client/src/client/impl/Command.cs [hunk 3]
```diff
@@ -68,18 +66,22 @@ static Command()
...
         }
 
         public Command(MethodBase method, ContentHeaderBase header, byte[] body)
         {
             Method = method;
-            Header = header;
-            m_body0 = body;
-            m_bodyN = null;
+            Header = header;
+            if (body != null)
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Command.cs]: Please use curly braces for single-statement conditionals. That appears to be the style in this codebase.

File: projects/client/RabbitMQ.Client/src/client/impl/Command.cs [hunk 4]
```diff
@@ -111,45 +113,13 @@ public static void CheckEmptyFrameSize()
...
-            }
-            else
-            {
-                if (m_bodyN == null)
-                {
-                    m_bodyN = new List<byte[]>();
-                }
-                m_bodyN.Add(fragment);
-            }
+            if(fragment !=null)
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Command.cs]: Please use a space after `if` as well as curly braces.

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 5]
```diff
@@ -363,7 +363,7 @@ IProtocol IConnection.Protocol
             table["platform"] = Encoding.UTF8.GetBytes(".NET");
             table["copyright"] = Encoding.UTF8.GetBytes("Copyright (c) 2007-2016 Pivotal Software, Inc.");
             table["information"] = Encoding.UTF8.GetBytes("Licensed under the MPL.  " +
-                                                          "See http://www.rabbitmq.com/");
+                                                          "See http://www.RabbitMQ.com/");
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: Please revert this change.

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 6]
```diff
@@ -1258,7 +1258,7 @@ void IDisposable.Dispose()
             }
             catch (OperationInterruptedException)
             {
-                // ignored, see rabbitmq/rabbitmq-dotnet-client#133
+                // ignored, see RabbitMQ/RabbitMQ-dotnet-client#133
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: Please revert this change.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 7]
```diff
@@ -1,218 +1,312 @@
...
+using Windows.Networking.Sockets;
+#else
+
+using System.Net.Sockets;
+
+#endif
+
+namespace RabbitMQ.Client.Impl
+{
+    public class HeaderWriteFrame : WriteFrame {
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Please open the curly brace on a new line.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 8]
```diff
@@ -1,218 +1,312 @@
...
+        public HeaderWriteFrame(int channel, ContentHeaderBase header, int bodyLength) : base(FrameType.FrameHeader, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            writer.Write((ushort)header.ProtocolClassId);
+            header.WriteTo(writer, (ulong)bodyLength);
+        }
+    }
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Please include a blank line here and in similar places.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 9]
```diff
@@ -1,218 +1,312 @@
...
-            FinishWriting();
-            writer.Write((byte)Type);
-            writer.Write((ushort)Channel);
-            writer.Write((uint)Payload.Length);
-            writer.Write(Payload);
-            writer.Write((byte)Constants.FrameEnd);
-        }
-    }
-}
+// This source code is dual-licensed under the Apache License, version
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Why is the diff viewer showing a difference here?

File: projects/client/RabbitMQ.Client/src/client/impl/ModelBase.cs [hunk 10]
```diff
@@ -1,1567 +1,1571 @@
...
-            public ConnectionSecureOrTune m_result;
-        }
-
-        public class QueueDeclareRpcContinuation : SimpleBlockingRpcContinuation
-        {
-            public QueueDeclareOk m_result;
-        }
-    }
-}
+// This source code is dual-licensed under the Apache License, version
```
Comment (lukebakken) [inline projects/client/RabbitMQ.Client/src/client/impl/ModelBase.cs]: This file should not be showing this many differences. Is there an issue with newlines?

**Comment by bording** [comment]
@YulerB I wanted to take a look at this PR, but there appear to be a bunch of changes that are unrelated to fixing lock contention. For example, lots of things have had their accessibility changed, which is a breaking change. Since this project is following SemVer now, that would mean this would require a 6.0 release of client, and I know the maintainers aren't going to be interested in doing that any time soon.

**Comment by michaelklishin** [comment]
@bording is right. We can accept minor breaking changes for a `5.1.0` but reducing lock contention should be possible with none or an minimum of breaking API changes.

It would also be interesting to see some benchmarks and other metrics that demonstrate the effect of this change.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 11]
```diff
@@ -1,321 +1,304 @@
...
+                nbw.Write((byte)Endpoint.Protocol.Revision);
+            }
+            else
+            {
+                nbw.Write(one);
+                nbw.Write(one);
+                nbw.Write((byte)Endpoint.Protocol.MajorVersion);
+                nbw.Write((byte)Endpoint.Protocol.MinorVersion);
+            }
+            m_writer.Write(ms.ToArray());
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Entire frame is written as one unit to the stream, thus removing the need to worry about frame interleaving.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 12]
```diff
@@ -1,321 +1,304 @@
...
+        }
+
+        public void WriteFrame(WriteFrame frame)
+        {
+            var ms = new MemoryStream();
+            var nbw = new NetworkBinaryWriter(ms);
+            frame.WriteTo(nbw);
+            m_socket.Client.Poll(m_writeableStateTimeout, SelectMode.SelectWrite);
+            m_writer.Write(ms.ToArray());
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Entire frame is written as one unit to the stream, thus removing the need to worry about frame interleaving.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 13]
```diff
@@ -1,321 +1,304 @@
...
+        }
+
+        public void WriteFrameSet(IList<WriteFrame> frames)
+        {
+            var ms = new MemoryStream();
+            var nbw = new NetworkBinaryWriter(ms);
+            foreach (var f in frames) f.WriteTo(nbw);
+            m_socket.Client.Poll(m_writeableStateTimeout, SelectMode.SelectWrite);
+            m_writer.Write(ms.ToArray());
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Entire frame is written as one unit to the stream, thus removing the need to worry about frame interleaving.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 14]
```diff
@@ -1,321 +1,304 @@
...
-            {
-                m_socket.Client.Poll(m_writeableStateTimeout, SelectMode.SelectWrite);
-                frame.WriteTo(m_writer);
-                m_writer.Flush();
-            }
-        }
-
-        public void WriteFrameSet(IList<Frame> frames)
-        {
-            lock (m_writer)
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Removed this lock
Frame interleaving is fixed by writing the entire frame as one unit.
The lock for the flush is not required when not using BufferedStream.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 15]
```diff
@@ -1,321 +1,304 @@
...
-                {
-                    f.WriteTo(m_writer);
-                }
-                m_writer.Flush();
-            }
-        }
-
-        public void Flush()
-        {
-            lock (m_writer)
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Removed this lock and flush, since we don't require flush when not using a BufferedStream.

File: projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs [hunk 16]
```diff
@@ -1,321 +1,304 @@
...
-        {
-            lock (m_reader)
-            {
-                return Frame.ReadFrom(m_reader);
-            }
-        }
-
-        public void SendHeader()
-        {
-            lock (m_writer)
```
Comment (YulerB) [inline projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs]: Removed this lock and flush, since we don't require flush when not using a BufferedStream and we don't require the lock on write because we will put the full frame on the stream as one unit.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs (hunk 9 above)
Comment (YulerB) [reply projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: The delta between old and new has too many changes to allow the compare tool to provide a true comparison,  the frame class has been extended with many frame super classes.

Also the base frame class data is now immutable.

**Comment by YulerB** [comment]
Made some changes, can we check again if there are breaking changes?

**Comment by sigiesec** [comment]
Could you please revert the whitespace changes? e.g. Command.cs and SocketFrameHandler.cs show changes in every line.

**Comment by michaelklishin** [comment]
I'd like to clarify a couple of things here before we continue reviewing this PR. Most changes here do not affect the public API: end user apps or even libraries such as NServiceBus never have to worry about frame serialization.

I do think that there can be subtle effects of this change that can only be revealed by a certain concurrent workload. Therefore I suggest that we consider this for a `5.1.0` release which will incubate for some time.

**Comment by michaelklishin** [comment]
@YulerB I believe the following should fix the line ending situation in `ModelBase.cs`:

 * Configure git to set [core.autocrlf to true]()
 * Rebase locally with --`whitespace=fix` [as explained in this thread](https://stackoverflow.com/questions/20977707/undo-line-ending-changes-in-git)
 * Force push to your repo: I expect that the PR will be updated, if not we can always submit a new one or pull manually on the command line

Alternatively you can switch line endings to UNIX using your IDE or editor of choice, as [core.autocrlf set to true has downsides as well](https://stackoverflow.com/questions/2825428/why-should-i-use-core-autocrlf-true-in-git). Note that you can switch `core.autocrlf` to `true` temporarily and then back.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs (hunk 9 above)
Comment (michaelklishin) [reply projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: It's actually easy to see what the difference is:

```
git log -p -- projects/client/RabbitMQ.Client/src/client/impl/Frame.cs
```

Existing line endings are replaced by Windows ones.

**Comment by adamralph** [comment]
It's tangential to this PR but, rather than relying on everyone setting `autocrlf` correctly, it would be better to put a [gitattributes file](https://git-scm.com/docs/gitattributes) in the repo, which controls those settings for everyone.

**Comment by YulerB** [comment]
If anyone knows how to fix that line ending issue with VS2017 installed, please go ahead.
@vendre21  do you have VS2017 to make the change?

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 17]
```diff
@@ -1,218 +1,327 @@
...
+            return base.ToString() + string.Format("(type={0}, channel={1}, {2} bytes of payload)",
+                Type,
+                Channel,
+                Payload == null
+                    ? "(null)"
+                    : Payload.Length.ToString());
+        }
+    }
+    
+    public class WriteFrame : Frame
```
Comment (michaelklishin) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Can we please rename this to `WriteableFrame` or `OutgoingFrame`? There are no "write frames" in the protocol (that would be different from "read frames") plus we have methods that are named `WriteFrame` and it gets fairly confusing even with a decent IDE.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 18]
```diff
@@ -1,218 +1,327 @@
...
+            
+            var argWriter = new MethodArgumentWriter(writer);
+            
+            method.WriteArgumentsTo(argWriter);
+            
+            argWriter.Flush();
+        }
+    }
+    
+    public class EmptyWriteFrame : WriteFrame
```
Comment (michaelklishin) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Same as below, I'd use `EmptyWriteableFrame` or `EmptyOutgoingFrame`.

File: projects/client/RabbitMQ.Client/src/client/impl/Frame.cs [hunk 19]
```diff
@@ -1,218 +1,327 @@
...
+            
+            writer.Write((byte)Type);
+            writer.Write((ushort)Channel);
+            writer.Write((uint)payload.Length);
+            writer.Write(payload);
+            writer.Write((byte)Constants.FrameEnd);
+        }
+    }
+    
+    public class ReadFrame : Frame
```
Comment (michaelklishin) [inline projects/client/RabbitMQ.Client/src/client/impl/Frame.cs]: Can we please rename this to `ReadableFrame` or `IncomingFrame`? There are no "write frames" in the protocol (that would be different from "read frames").

**Comment by michaelklishin** [comment]
This client's tests pass and the idea behind this change seems sound to me:

 * Frames and frame sets (for frames that carry content and need, ahem, content framing) are written in a single operation to a memory-backed stream
 * `BufferedStream` is no longer used (it was not efficiently used for writes anyway)
 * The above seeks to guarantee that socket stream writes cannot be interleaved in a way that will confuse RabbitMQ protocol parser
 * As such, locking is no longer necessary

@bording @danielmarbach do you see any flaws in the above? any incorrect assumptions I may be overlooking?

**Comment by michaelklishin** [comment]
@YulerB thanks again for your work on this. This is promising but I'm afraid cannot be merged as is: there are too many line ending changes and they make it very hard to notice if there are any changes we are not noticing.

It may be possible to fix them, rebase and force push this PR. Alternatively a new PR without line endings can be submitted and we will review and test it instead. I'll leave this one open in hope that a force push would suffice.

**Comment by michaelklishin** [comment]
@adamralph I agree. We'll do it separately.

**Comment by YulerB** [comment]
Do you want me to leave the PR open for you guys or close?

**Comment by lukebakken** [comment]
@adamralph - the `.gitattributes` file is [already present](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/.gitattributes). It should have prevented the newline issues in this PR. I'm not sure why it didn't.

**Comment by michaelklishin** [comment]
@YulerB I'd prefer to keep this one going because we have a review history here but we can do either.

**Comment by YulerB** [comment]
No worries, ill leave it open for now, our local version has further improvements, may try another PR in the future...
I think I'll request vs2017 first.

**Comment by vendre21** [comment]
I tried to change the Frame.cs file, can you check the line endings on that file please.

**Comment by michaelklishin** [comment]
@vendre21 accorindg to `git diff master..HEAD` the commits that changed line endings of every line in `Frame.cs` and `Command.cs` are still there, so the diff both on the command line and in this PR still includes them.

**Comment by vendre21** [comment]
Thank you @michaelklishin. I think I could fix the line endings this time.

File: projects/client/RabbitMQ.Client/src/client/impl/Command.cs [hunk 20]
```diff
@@ -164,54 +134,27 @@ public void Transmit(int channelNumber, Connection connection)
             }
         }
 
-        public void TransmitAsSingleFrame(int channelNumber, Connection connection)
+        private void TransmitAsSingleFrame(int channelNumber, Connection connection)
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Command.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Command.cs [hunk 21]
```diff
@@ -164,54 +134,27 @@ public void Transmit(int channelNumber, Connection connection)
...
-            writer.Write((ushort)Method.ProtocolMethodId);
-            var argWriter = new MethodArgumentWriter(writer);
-            Method.WriteArgumentsTo(argWriter);
-            argWriter.Flush();
-            connection.WriteFrame(frame);
+            connection.WriteFrame(new MethodWriteFrame(channelNumber, Method));
         }
 
-        public void TransmitAsFrameSet(int channelNumber, Connection connection)
+        private void TransmitAsFrameSet(int channelNumber, Connection connection)
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Command.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 22]
```diff
@@ -273,7 +273,7 @@ public ushort ChannelMax
         public IDictionary<string, object> ClientProperties
         {
             get { return m_clientProperties; }
-            set { m_clientProperties = value; }
+            private set { m_clientProperties = value; }
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 23]
```diff
@@ -339,7 +339,7 @@ public int RemotePort
             get { return m_frameHandler.RemotePort; }
         }
 
-        public IDictionary<string, object> ServerProperties { get; set; }
+        public IDictionary<string, object> ServerProperties { get; private set; }
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 24]
```diff
@@ -465,7 +465,7 @@ public void Close(ShutdownEventArgs reason, bool abort, int timeout)
         ///<remarks>
         /// Loop only used while quiescing. Use only to cleanly close connection
         ///</remarks>
-        public void ClosingLoop()
+        private void ClosingLoop()
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 25]
```diff
@@ -533,7 +533,7 @@ public void EnsureIsOpen()
         }
 
         // Only call at the end of the Mainloop or HeartbeatLoop
-        public void FinishClose()
+        private void FinishClose()
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: This seems to be a public API breaking change

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 26]
```diff
@@ -561,13 +561,13 @@ public void HandleApplicationSuspend(object sender, SuspendingEventArgs suspendi
         /// We need to close the socket, otherwise attempting to unload the domain
         /// could cause a CannotUnloadAppDomainException
         /// </remarks>
-        public void HandleDomainUnload(object sender, EventArgs ea)
+        private void HandleDomainUnload(object sender, EventArgs ea)
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: This seems to be a public API breaking change as well as following

File: projects/client/RabbitMQ.Client/src/client/impl/TcpClientAdapter.cs [hunk 27]
```diff
@@ -32,7 +32,7 @@ public TcpClientAdapter(Socket socket)
                 throw new ArgumentException("No ip address could be resolved for " + host);
             }
             #if CORECLR
-            await sock.ConnectAsync(ep, port).ConfigureAwait(false);
```
Comment (danielmarbach) [inline projects/client/RabbitMQ.Client/src/client/impl/TcpClientAdapter.cs]: ConfigureAwait is required

**Comment by YulerB** [comment]
Since we cannot make changes on the main branch, we can only make changes from a fork.
Therefore all the changes in the fork were part of the PR. 

I like the split of the responsibilities for generating frames and reading frames.

I believe @michaelklishin wants to work on this taking the best out of the PR and adding to the trunk.

We are considering other PR's, I'll need the rest of my team to create forks though, since looks like GitHub only allows one per user, may be wrong, we are new to github.

We have a change to the AsyncConsumerWorkService we would also like to include, but this is to simplify some code and to also ensure there are no locking issues in the current code (A little clue, BlockingCollection<Work>).  I believe @vendre21  will submit soon.

We also see allot of casting in the application, and have extensions to the networkbinaryreader/networkbinarywriter to avoid doing casts.

We also see allot of stuff like (byte) 'S' which really should be a constant byte S = 68;, rather than 2 opcodes to allocate and cast each method call, for example WireFormatter.

File: projects/client/RabbitMQ.Client/src/client/impl/Command.cs (hunk 20 above)
Comment (michaelklishin) [reply projects/client/RabbitMQ.Client/src/client/impl/Command.cs]: This method is never called by end user apps or libraries that depend on this client. See https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/350#issuecomment-329123265. So we can undo it but if someone's code depends on this, they are doing something really risky to begin with.

**Comment by lukebakken** [comment]
@YulerB - please limit each PR to the minimal set of changes necessary for a particular feature, bug fix, or refactor. For instance, modifying methods to be `private` should not be part of this PR. Making micro-optimizations like removing `(byte)S` should be a separate PR.

It is much easier to review and validate several separate pull requests instead of one monolithic one. If you'd like, close this PR and break it up into several smaller ones. There is no "cost" to closing PRs and opening new ones.

In addition, please try to adhere to the coding style of the project which will reduce the number of review comments made.

If you would like your team to work on the same fork, then I suggest forking this project to an organization that your team can be made members of. As an alternative, one team member can fork this project to their individual GitHub account and then via the repository's settings, add collaborators to the fork.

Thanks a lot for your contributions.

**Comment by lukebakken** [comment]
@YulerB -

Please disregard the comment I made earlier, I was looking at a stale version of your PR.

> We are considering other PR's, I'll need the rest of my team to create forks though, since looks like GitHub only allows one per user, may be wrong, we are new to github.

You can create an organiziation to which to make your fork, and then add your team to that organization. As an alternative, one person can fork this repo and via the fork's settings (`https://github.com/USERNAME/rabbitmq-dotnet-client/settings/collaboration`) add collaborators.

Thanks!!

**Comment by michaelklishin** [comment]
The diff is much smaller now, thank you, @YulerB.

Some of the unrelated eg. member visibility changes are not ideal from PR hygiene perspective but I can see why they were done: this client has a complex history and a fair amount of curious (as far as C# coding style goes) decisions that go back all the way to 2007. So it's attracting drive-by changes.

**Comment by michaelklishin** [comment]
I QA'ed it yesterday and so far it looks promising. I'm going to look into review our concurrent publishing tests and see if we may need to add a couple more.

@bording @danielmarbach it would be great to have a review and some NServiceBus-specific testing from you folks :)

**Comment by danielmarbach** [comment]
I'm not an expert when it comes to buffered streams. Based on the limited knowledge I have around it I know they are overrated when it comes to FileStreams since the file stream itself has already built in buffering. I don't know about the network streams. That is where my knowledge ends. I do think though this has to be taken into account and at least verified. Since what might look like an optimization for small and high number of messages could turn around as a major performance drawback when considering larger message payloads if the BufferedStream was purposefully introduced for those scenarios. If that scenario can be rules out I'd say it is worth proceeding with the reduction of the lock contention since we know based on the history in this client how drastic this can improve performance.

**Comment by michaelklishin** [comment]
I'm adding concurrent connection access tests in preparation for further QA of this pull request.

**Comment by pivotal-issuemaster** [comment]
@YulerB Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-dotnet-client&pullRequestId=350)!

[Click here](https://cla.pivotal.io/sync/rabbitmq?repositoryId=rabbitmq/rabbitmq-dotnet-client&pullRequestId=350) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by danielmarbach** [comment]
Can you also add payload tests with several payload sizes?

**Comment by michaelklishin** [comment]
I pushed this branch together with some extra tests in master as https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/354. Let's continue there.

**Comment by michaelklishin** [comment]
@danielmarbach good idea. I will do that on the #354 branch.

**Comment by pivotal-issuemaster** [comment]
@YulerB Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-dotnet-client&pullRequestId=350)!

**Comment by michaelklishin** [comment]
@YulerB unfortunately, we had to reintroduce synchronization for non-TLS socket writes in #683 as #681 highlighted that it was not a safe thing to do on some platforms :( The change will ship in `6.0`.
